### PR TITLE
Implement Typestate Builder Pattern

### LINKED ISSUE #10: Implement Typestate Builder Pattern
**Is your feature request related to a problem? Please describe.**
Currently, requests are validated at runtime, before they are sent out to the server.
It would be better if the requests are validated at compiletime, so that developers using our library can create more robust code with as little runtime errors as possible, creating a smoother developing experience in the end.

**Describe the solution you'd like**
Thomas suggested using the (TypeState) builder pattern, so that only requests that are complete have an implemented build() method.

### PR DESCRIPTION
**Pull request by ejorritsma**
### :sparkles: What kind of change does this PR introduce? (Bug fix, feature, docs update...)
A rewrite of the way request building works. This introduces the advantage of validation of requests at compile time.

### :arrow_heading_down: What is the current behavior?
Currently, requests are validated at runtime, which may lead to unexpected behavior.

### :new: What is the new behavior (if this is a feature change)?
Now, requests are validated at compile time.

### :boom: Does this PR introduce a breaking change?
Yes. However, since we are not running in production yet, I think we will be ok.

### :bug: Recommendations for testing
Run the tests, which have been unchanged.

### :memo: Links to relevant issues/docs
https://github.com/Baseflow/stellar-rust-sdk/issues/10

### :thinking: Checklist before submitting

- [x] All projects build
- [x] Follows style guide lines 
- [x] Relevant documentation was updated
- [x] Rebased onto current main

### CODE REVIEW

File: src/assets/all_assets_request.rs [hunk 1]
```diff
@@ -3,9 +3,9 @@ use crate::models::Request;
 use super::super::Order;
```
Comment (tluijken) [inline src/assets/all_assets_request.rs]: Unrelated to this PR, but we could just use crate::Order, in stead of traversing up `super::super`.

File: src/assets/all_assets_request.rs [hunk 2]
```diff
@@ -107,9 +73,15 @@ impl AllAssetsRequest {
     /// # Returns
     /// The request object
     /// [AllAssetsRequest](struct.AllAssetsRequest.html)
-    pub fn set_asset_code(&mut self, asset_code: &str) -> &mut Self {
-        self.asset_code = Some(asset_code.to_owned());
-        self
+    pub fn set_asset_code(self, asset_code: &str) -> Result<AllAssetsRequest, String> {
+        if asset_code.len() > 12 {
+            return Err("asset_code must be 12 characters or less".to_string());
+        }
```
Comment (tluijken) [inline src/assets/all_assets_request.rs:65]: These are still runtime errors. We could introduce an 'InvalidAssetRequest' state and return this if the asset_code > 12.

File: src/assets/all_assets_request.rs [hunk 3]
```diff
@@ -118,9 +90,15 @@ impl AllAssetsRequest {
     /// # Returns
     /// The request object
     /// [AllAssetsRequest](struct.AllAssetsRequest.html)
-    pub fn set_asset_issuer(&mut self, asset_issuer: &str) -> &mut Self {
-        self.asset_issuer = Some(asset_issuer.to_owned());
-        self
+    pub fn set_asset_issuer(self, asset_issuer: &str) -> Result<AllAssetsRequest, String> {
+        if asset_issuer.len() != 56 {
+            return Err("asset_issuer must be 56 characters".to_string());
+        }
```
Comment (tluijken) [inline src/assets/all_assets_request.rs:82]: These are still runtime errors. We could introduce an 'InvalidAssetRequest' state and return this if the asset_issuer > 12.

File: src/assets/all_assets_request.rs [hunk 4]
```diff
@@ -118,9 +90,15 @@ impl AllAssetsRequest {
     /// # Returns
     /// The request object
     /// [AllAssetsRequest](struct.AllAssetsRequest.html)
-    pub fn set_asset_issuer(&mut self, asset_issuer: &str) -> &mut Self {
-        self.asset_issuer = Some(asset_issuer.to_owned());
-        self
+    pub fn set_asset_issuer(self, asset_issuer: &str) -> Result<AllAssetsRequest, String> {
```
Comment (tluijken) [inline src/assets/all_assets_request.rs:79]: We should add the boundaries to the documentation, so the developers know if they mess up, why their code doesn't compile.

File: src/assets/all_assets_request.rs [hunk 5]
```diff
@@ -107,9 +73,15 @@ impl AllAssetsRequest {
     /// # Returns
     /// The request object
     /// [AllAssetsRequest](struct.AllAssetsRequest.html)
-    pub fn set_asset_code(&mut self, asset_code: &str) -> &mut Self {
-        self.asset_code = Some(asset_code.to_owned());
-        self
+    pub fn set_asset_code(self, asset_code: &str) -> Result<AllAssetsRequest, String> {
```
Comment (tluijken) [inline src/assets/all_assets_request.rs:62]: We should add the boundaries to the documentation, so the developers know if they mess up, why their code doesn't compile.

File: src/assets/all_assets_request.rs [hunk 6]
```diff
@@ -129,9 +107,15 @@ impl AllAssetsRequest {
     /// # Returns
     /// The request object
     /// [AllAssetsRequest](struct.AllAssetsRequest.html)
-    pub fn set_cursor(&mut self, cursor: u32) -> &mut Self {
-        self.cursor = Some(cursor);
-        self
+    pub fn set_cursor(self, cursor: u32) -> Result<AllAssetsRequest, String> {
```
Comment (tluijken) [inline src/assets/all_assets_request.rs:96]: We should add the boundaries to the documentation, so the developers know if they mess up, why their code doesn't compile.

File: src/assets/all_assets_request.rs [hunk 7]
```diff
@@ -129,9 +107,15 @@ impl AllAssetsRequest {
...
+        if cursor < 1 {
+            return Err("cursor must be greater than or equal to 1".to_string());
```
Comment (tluijken) [inline src/assets/all_assets_request.rs:98]: These are still runtime errors. We could introduce an 'InvalidAssetRequest' state and return this if the cursor < 1. 
By the way, what happens if this is 0? Is this really an invalid use case? The docs do not describe a limitation for this.

File: src/assets/all_assets_request.rs [hunk 8]
```diff
@@ -140,9 +124,15 @@ impl AllAssetsRequest {
     /// # Returns
     /// The request object
     /// [AllAssetsRequest](struct.AllAssetsRequest.html)
-    pub fn set_limit(&mut self, limit: u32) -> &mut Self {
-        self.limit = Some(limit);
-        self
+    pub fn set_limit(self, limit: u32) -> Result<AllAssetsRequest, String> {
```
Comment (tluijken) [inline src/assets/all_assets_request.rs:113]: We should set the limit type to u8, which only allows a maximum of 255. This is still out of bounds, but at least I cannot add a large value like 10.000. This narrows down the fail rate.

File: src/assets/all_assets_request.rs [hunk 9]
```diff
@@ -140,9 +124,15 @@ impl AllAssetsRequest {
...
+        if limit < 1 || limit > 200 {
+            return Err("limit must be between 1 and 200".to_string());
+        }
```
Comment (tluijken) [inline src/assets/all_assets_request.rs:116]: These are still runtime errors. We could introduce an 'InvalidAssetRequest' state and return this if the limit < 1 or > 200.

File: src/assets/all_assets_request.rs (hunk 8 above)
Comment (tluijken) [inline src/assets/all_assets_request.rs:113]: We should add the limit boundaries to the documentation, so developers know why their code no longer compiles if it does so.

File: src/claimable_balances/all_claimable_balances_request.rs [hunk 10]
```diff
@@ -28,71 +29,9 @@ pub struct AllClaimableBalancesRequest {
...
-
-        if let Some(claimant) = &self.claimant {
-            let is_valid = is_public_key(claimant);
-            if is_valid.is_err() {
-                return Err(is_valid.unwrap_err());
-            }
-        }
-
-        Ok(())
+impl AllClaimableBalancesRequest {
```
Comment (tluijken) [inline src/claimable_balances/all_claimable_balances_request.rs]: A lot of the comments from the all assets request also apply here. There should be an `InvalidClaimableBalancesRequest` state, which does not allow for an empty request without any parameters.
Don't forget to add boundaries to the documentation, and check the type for limit, to trim it down to the smallest possible value that matches the limits of the request.

File: src/claimable_balances/single_claimable_balance_request.rs [hunk 11]
```diff
@@ -1,30 +1,43 @@
...
+/// [More Details](https://developers.stellar.org/api/horizon/resources/retrieve-a-claimable-balance) "Single Claimable Balance")
+#[derive(Default)]
+pub struct SingleClaimableBalanceRequest<I> {
+    claimable_balance_id: I,
+}
+
+impl SingleClaimableBalanceRequest<NoClaimableBalanceId> {
+    pub fn new() -> Self {
+        SingleClaimableBalanceRequest::default()
+    }
```
Comment (tluijken) [inline src/claimable_balances/single_claimable_balance_request.rs]: We should add some documentation. It seems obvious, but that does this actually do?

File: src/accounts/single_account_request.rs [hunk 12]
```diff
@@ -1,26 +1,50 @@
 use crate::models::{is_public_key, Request};
 
-/// SingleAccountRequest is the request for the /accounts enpoint to get a single account.
-/// [More Details](https://www.stellar.org/developers/horizon/reference/endpoints/accounts.html "Accounts")
-pub struct SingleAccountRequest {
+#[derive(Default, Clone)]
+pub struct AccountId(String);
+#[derive(Default, Clone)]
+pub struct NoAccountId;
```
Comment (tluijken) [inline src/accounts/single_account_request.rs:6]: We should add some documentation on the context of each of these states. What do they mean,?

File: src/claimable_balances/single_claimable_balance_request.rs [hunk 13]
```diff
@@ -1,30 +1,43 @@
 use crate::models::*;
 
+
+#[derive(Default, Clone)]
+pub struct ClaimableBalanceId(String);
+#[derive(Default, Clone)]
+pub struct NoClaimableBalanceId;
+
```
Comment (tluijken) [inline src/claimable_balances/single_claimable_balance_request.rs:7]: We should add some documentation on the context of each of these states. What do they mean?

File: src/claimable_balances/single_claimable_balance_request.rs [hunk 14]
```diff
@@ -1,30 +1,43 @@
...
-    fn new() -> Self {
-        SingleClaimableBalanceRequest {
-            claimable_balance_id: None,
+impl<I> SingleClaimableBalanceRequest<I> {
+    pub fn set_claimable_balance_id(
+        self,
+        claimable_balance_id: String,
+    ) -> SingleClaimableBalanceRequest<ClaimableBalanceId> {
+        SingleClaimableBalanceRequest { 
+            claimable_balance_id: ClaimableBalanceId(claimable_balance_id)
```
Comment (tluijken) [inline src/claimable_balances/single_claimable_balance_request.rs]: What would be really awesome is that we could implement the `set_claimable_balance_id` on the implementation of `SingleClaimableBalanceRequest<NoClaimableBalanceId>`. Once the balance is set, there is way to set it again, as this function if no longer implemented for `SingleClaimableBalanceRequest<ClaimableBalanceId>`

This trick might apply to other request builders too.

File: src/claimable_balances/single_claimable_balance_request.rs [hunk 15]
```diff
@@ -1,30 +1,43 @@
...
+        SingleClaimableBalanceRequest::default()
+    }
 }
 
-impl Request for SingleClaimableBalanceRequest {
-    fn new() -> Self {
-        SingleClaimableBalanceRequest {
-            claimable_balance_id: None,
+impl<I> SingleClaimableBalanceRequest<I> {
+    pub fn set_claimable_balance_id(
```
Comment (tluijken) [inline src/claimable_balances/single_claimable_balance_request.rs:20]: Please add some documentation on what this does, if there are any limitations on the input, and what the output will be.

File: src/horizon_client/horizon_client.rs [hunk 16]
```diff
@@ -60,7 +59,7 @@ impl HorizonClient {
     /// [GET /accounts/{account_id}](https://www.stellar.org/developers/horizon/reference/endpoints/accounts-single.html)
     pub async fn get_single_account(
         &self,
-        request: &SingleAccountRequest,
+        request: &SingleAccountRequest<AccountId>,
```
Comment (tluijken) [inline src/horizon_client/horizon_client.rs:62]: Why is there a difference between the `&impl ValidAccountsRequest` and `&SingleAccountRequest<AccountId>` in both signatures (looking at `get_account_list`)?

File: src/horizon_client/horizon_client.rs [hunk 17]
```diff
@@ -243,10 +236,9 @@ mod tests {
             HorizonClient::new("https://horizon-testnet.stellar.org".to_string()).unwrap();
 
         // construct request
-        let mut accounts_request = AccountsRequest::new();
-        accounts_request
-            .set_signer("GDQJUTQYK2MQX2VGDR2FYWLIYAQIEGXTQVTFEMGH2BEWFG4BRUY4CKI7")
-            .set_limit(10);
+        let accounts_request = AccountsRequest::new()
```
Comment (tluijken) [inline src/horizon_client/horizon_client.rs:239]: We can also add unit tests for invalid states by checking that they result in the correct type:
`.is::<T>()`

File: src/ledgers/ledgers_request.rs [hunk 18]
```diff
@@ -79,17 +58,38 @@ impl LedgersRequest {
     /// # Returns
     /// The request object
     /// [AllLedgersRequest](struct.AllLedgersRequest.html)
-    pub fn set_cursor(&mut self, cursor: u32) {
-        self.cursor = Some(cursor);
+    pub fn set_cursor(self, cursor: u32) -> Result<LedgersRequest, String>{
+        if cursor < 1 {
+            return Err("cursor must be greater than or equal to 1".to_string());
+        }
```
Comment (tluijken) [inline src/ledgers/ledgers_request.rs:27]: The same remarks as ealier. These are still runtime exceptions, but could return an `InvalidLedgerRequest` state. 
Please update the docs indicating what the boundaries are for the cursor.

File: src/ledgers/ledgers_request.rs [hunk 19]
```diff
@@ -79,17 +58,38 @@ impl LedgersRequest {
...
     /// * `limit` - The limit
     /// # Returns
     /// The request object
     /// [AllAssetsRequest](struct.AllAssetsRequest.html)
-    pub fn set_limit(&mut self, limit: u32) {
-        self.limit = Some(limit);
+    pub fn set_limit(self, limit: u32) -> Result<LedgersRequest, String> {
+        if limit < 1 || limit > 200 {
+            return Err("limit must be between 1 and 200".to_string());
+        }
```
Comment (tluijken) [inline src/ledgers/ledgers_request.rs:45]: The same remarks as ealier. These are still runtime exceptions, but could return an `InvalidLedgerRequest` state. 
Please update the docs indicating what the boundaries are for the cursor.
Also, let's size limit down to a u8.

File: src/ledgers/single_ledger_request.rs [hunk 20]
```diff
@@ -1,32 +1,53 @@
...
+    /// # Returns
+    /// The request object
+    /// [SingleLedgerRequest](struct.SingleLedgerRequest.html)
+    pub fn set_sequence(
+        self,
+        sequence: u32,
+    ) -> Result<SingleLedgerRequest<Sequence>, String> {
+        if sequence < 1 {
+            return Err("sequence must be greater than or equal to 1".to_string());
+        }
```
Comment (tluijken) [inline src/ledgers/single_ledger_request.rs]: Same as before. We should implement this for a `NoSequence` variant and return a `Sequence` variant. We can set this only once. Try to remedy this runtime error with an `InvalidSequence` state.

File: src/models/mod.rs [hunk 21]
```diff
@@ -1,11 +1,11 @@
...
-    /// Creates a new request object
-    /// # Returns
-    /// A new request object
-    /// [Request](trait.Request.html)
-    fn new() -> Self;
+    // /// Creates a new request object
+    // /// # Returns
+    // /// A new request object
+    // /// [Request](trait.Request.html)
+    // fn new() -> Self;
```
Comment (tluijken) [inline src/models/mod.rs]: Let's remove this

File: src/models/mod.rs [hunk 22]
```diff
@@ -21,12 +21,12 @@ pub trait Request {
...
-    /// Returns Ok(()) if the request is valid
-    /// This method is called before the request is sent
-    /// to the server
-    fn validate(&self) -> Result<(), String>;
+    // /// Validate the request
+    // /// Returns an error if the request is invalid
+    // /// Returns Ok(()) if the request is valid
+    // /// This method is called before the request is sent
+    // /// to the server
+    // fn validate(&self) -> Result<(), String>;
```
Comment (tluijken) [inline src/models/mod.rs]: Let's remove this

**Review by tluijken — CHANGES_REQUESTED** [review]
Great work so far! I have a lot of comments, but don't worry, it's more of the same for a lot of this. I'll check in with you personally do elaborate the high level changes.

File: src/horizon_client/horizon_client.rs (hunk 16 above)
Comment (ejorritsma) [reply src/horizon_client/horizon_client.rs:62]: not sure I understand correctly.
The get_account_list() can accept a variety of AccountsRequest objects.
The get_single_account() can only accept a single variety of the SingleAccountRequest, namely one with an AccountId. 

Does this make any sense? Maybe we can pick this up in an `validation improvement branch`, is this related to your suggestions for a Valid and Invalid state for all request types? :)

File: src/assets/all_assets_request.rs (hunk 2 above)
Comment (ejorritsma) [reply src/assets/all_assets_request.rs:65]: moved comment to https://github.com/Baseflow/stellar-rust-sdk/issues/8

File: src/assets/all_assets_request.rs (hunk 3 above)
Comment (ejorritsma) [reply src/assets/all_assets_request.rs:82]: moved comment to https://github.com/Baseflow/stellar-rust-sdk/issues/8

File: src/assets/all_assets_request.rs (hunk 4 above)
Comment (ejorritsma) [reply src/assets/all_assets_request.rs:79]: Moved comment to https://github.com/Baseflow/stellar-rust-sdk/issues/4

File: src/assets/all_assets_request.rs (hunk 5 above)
Comment (ejorritsma) [reply src/assets/all_assets_request.rs:62]: Moved to https://github.com/Baseflow/stellar-rust-sdk/issues/4

File: src/assets/all_assets_request.rs (hunk 6 above)
Comment (ejorritsma) [reply src/assets/all_assets_request.rs:96]: Moved to https://github.com/Baseflow/stellar-rust-sdk/issues/4

File: src/assets/all_assets_request.rs (hunk 7 above)
Comment (ejorritsma) [reply src/assets/all_assets_request.rs:98]: Moved to https://github.com/Baseflow/stellar-rust-sdk/issues/8

File: src/assets/all_assets_request.rs (hunk 8 above)
Comment (ejorritsma) [reply src/assets/all_assets_request.rs:113]: Moved to https://github.com/Baseflow/stellar-rust-sdk/issues/8

File: src/assets/all_assets_request.rs (hunk 9 above)
Comment (ejorritsma) [reply src/assets/all_assets_request.rs:116]: Moved to https://github.com/Baseflow/stellar-rust-sdk/issues/8

File: src/assets/all_assets_request.rs (hunk 8 above)
Comment (ejorritsma) [reply src/assets/all_assets_request.rs:113]: Moved to https://github.com/Baseflow/stellar-rust-sdk/issues/4

File: src/claimable_balances/all_claimable_balances_request.rs (hunk 10 above)
Comment (ejorritsma) [reply src/claimable_balances/all_claimable_balances_request.rs]: Moved to https://github.com/Baseflow/stellar-rust-sdk/issues/4 and https://github.com/Baseflow/stellar-rust-sdk/issues/8

File: src/claimable_balances/single_claimable_balance_request.rs (hunk 11 above)
Comment (ejorritsma) [reply src/claimable_balances/single_claimable_balance_request.rs]: Moved to https://github.com/Baseflow/stellar-rust-sdk/issues/4

File: src/accounts/single_account_request.rs (hunk 12 above)
Comment (ejorritsma) [reply src/accounts/single_account_request.rs:6]: Moved to https://github.com/Baseflow/stellar-rust-sdk/issues/4

File: src/claimable_balances/single_claimable_balance_request.rs (hunk 13 above)
Comment (ejorritsma) [reply src/claimable_balances/single_claimable_balance_request.rs:7]: Moved to https://github.com/Baseflow/stellar-rust-sdk/issues/4

File: src/claimable_balances/single_claimable_balance_request.rs (hunk 15 above)
Comment (ejorritsma) [reply src/claimable_balances/single_claimable_balance_request.rs:20]: Moved to https://github.com/Baseflow/stellar-rust-sdk/issues/4

File: src/ledgers/ledgers_request.rs (hunk 18 above)
Comment (ejorritsma) [reply src/ledgers/ledgers_request.rs:27]: Moved to https://github.com/Baseflow/stellar-rust-sdk/issues/4 and https://github.com/Baseflow/stellar-rust-sdk/issues/8

File: src/ledgers/ledgers_request.rs (hunk 19 above)
Comment (ejorritsma) [reply src/ledgers/ledgers_request.rs:45]: Moved to https://github.com/Baseflow/stellar-rust-sdk/issues/4 and https://github.com/Baseflow/stellar-rust-sdk/issues/8

File: src/ledgers/single_ledger_request.rs (hunk 20 above)
Comment (ejorritsma) [reply src/ledgers/single_ledger_request.rs]: Moved to https://github.com/Baseflow/stellar-rust-sdk/issues/8

File: src/assets/all_assets_request.rs (hunk 1 above)
Comment (ejorritsma) [reply src/assets/all_assets_request.rs]: Just fixed this!

File: src/claimable_balances/single_claimable_balance_request.rs (hunk 14 above)
Comment (ejorritsma) [reply src/claimable_balances/single_claimable_balance_request.rs]: Made non-optional parameters immutable!

File: src/horizon_client/horizon_client.rs (hunk 17 above)
Comment (ejorritsma) [reply src/horizon_client/horizon_client.rs:239]: Cannot figure out how to do this, could you point me in the right direction? For example, I get the error
`no method named `is` found for struct `AccountsRequest` in the current scope`

When adding
`assert!(accounts_request.is::<AccountsRequest<NoSponsor, AccountsSigner, NoAsset, NoLiquidityPool>>());`

**Pull request merged**